zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxnFactory.java

```diff
@@ -420,7 +431,11 @@ public void operationComplete(Future<Channel> future) {
                         return;
                     }
 
-                    if (KeeperException.Code.OK != authProvider.handleAuthentication(cnxn, null)) {
+                    KeeperException.Code code = authProvider.handleAuthentication(cnxn, null);
+                    if (KeeperException.Code.OK != code) {
+                        if (KeeperException.Code.AUTHFAILED == code) {
```

Comment:
This additional check seems redundant. We are making a binary decision here. I don't think `handleAuthentication ` has a third return type other than ok/fail.